grails-app/services/org/pih/warehouse/product/ProductIdentifierService.groovy

```diff
@@ -0,0 +1,199 @@
+package org.pih.warehouse.product
+
+import java.util.regex.Matcher
+import java.util.regex.Pattern
+import org.apache.commons.lang.StringUtils
+
+import org.pih.warehouse.core.Constants
+import org.pih.warehouse.core.IdentifierGeneratorTypeCode
+import org.pih.warehouse.core.IdentifierService
+import org.pih.warehouse.core.identification.BlankIdentifierResolver
+import org.pih.warehouse.core.identification.IdentifierGeneratorParams
+
+// TODO: We should be able to move most of this logic into the identifier service. That way we can support sequential
+//       id generation more generally. We'd need an abstract method there for fetching the next sequence number to use,
+//       and a new ${sequence} config option to know where to put the sequence.
+class ProductIdentifierService extends IdentifierService implements BlankIdentifierResolver<Product> {
```

Comment:
A majority of this code is just copied out from IdentifierService and some other places. I couldn't get around to refactoring this as a part of this change because it has such a complex configuration.

See the TODO here for how we could refactor in the future.

Reply:
Yeah this one is though, perhaps we could get more tests for it? This is not a change request for now, but might be an entry point to bigger refactor of that part